rust-runtime/aws-smithy-http-server/src/plugin/pipeline.rs

```diff
@@ -0,0 +1,143 @@
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{filter_by_operation_name, PluginPipeline};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as MetricsPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+/// # struct CheckHealth;
+/// # impl CheckHealth { const NAME: &'static str = "MyName"; }
+///
+/// let pipeline = PluginPipeline::new(LoggingPlugin)
+///     .push(MetricsPlugin)
+///     .map(|current_pipeline| {
+///         // The logging and metrics plugins will not be applied to the `CheckHealth` operation.
+///         filter_by_operation_name(current_pipeline, |name| name != CheckHealth::NAME)
+///     })
+///     // The auth plugin will be applied to all operations
+///     .push(AuthPlugin);
+/// ```
+///
+/// ## Concatenating two plugin pipelines
+///
+/// `PluginPipeline` is a good way to bundle together multiple plugins, ensuring they are all
+/// registered in the correct order.
+///
+/// You can use the [`concat`](PluginPipeline::concat) to append, at once, all the plugins
+/// in another pipeline to the current pipeline:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{PluginPipeline, PluginStack};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as MetricsPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+///
+/// pub fn get_bundled_pipeline() -> PluginPipeline<PluginStack<LoggingPlugin, MetricsPlugin>> {
+///     PluginPipeline::new(LoggingPlugin).push(MetricsPlugin)
+/// }
+///
+/// let pipeline = PluginPipeline::new(AuthPlugin)
+///     .concat(get_bundled_pipeline());
+/// ```
+///
+/// ## Providing custom methods on `PluginPipeline`
+///
+/// You use an **extension trait** to add custom methods on `PluginPipeline`.
+///
+/// This is a simple example using `AuthPlugin`:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{PluginPipeline, PluginStack};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+///
+/// pub trait AuthPluginExt<CurrentPlugins> {
+///     fn with_auth(self) -> PluginPipeline<PluginStack<CurrentPlugins, AuthPlugin>>;
+/// }
+///
+/// impl<CurrentPlugins> AuthPluginExt<CurrentPlugins> for PluginPipeline<CurrentPlugins> {
+///     fn with_auth(self) -> PluginPipeline<PluginStack<CurrentPlugins, AuthPlugin>> {
+///         self.push(AuthPlugin)
+///     }
+/// }
+///
+/// let pipeline = PluginPipeline::new(LoggingPlugin)
+///     // Our custom method!
+///     .with_auth();
+/// ```
+pub struct PluginPipeline<P>(P);
+
+impl PluginPipeline<IdentityPlugin> {
+    pub fn empty() -> Self {
+        Self(IdentityPlugin)
+    }
+}
+
+impl<P> PluginPipeline<P> {
+    pub fn new(new_plugin: P) -> PluginPipeline<P> {
```

Comment:
You can use `Self` in all of these.

Reply:
I have updated `new` in https://github.com/awslabs/smithy-rs/pull/1971/commits/aa08685cffb518498ed1c3ba1539269c2ba33d32
The other methods cannot use `Self` since the type of the wrapped plugin changes.